### PR TITLE
Support index level allocation filtering for searchable snapshot index

### LINKED ISSUE #11468: Searchable snapshot supports index level allocation filter settings
**Is your feature request related to a problem? Please describe.**
Today, the searchable snapshot index depends on the `RemoteShardsBalancer` to carry out allocation operations. However, I have noticed that this balancer does not support allocation filter settings at the index level. It only considers cluster level settings when performing `moveShards`. I am curious as to why this setting is not supported.

**Describe the solution you'd like**
We can filter out the shards that cannot remain on the current node and prioritize moving them to eligible nodes before considering moving shards on excluded nodes.
```
List<ShardRouting> forceMoveShards = new ArrayList<>();
eligibleNodes.forEach(sourceNode -> {
    for (ShardRouting shardRouting : sourceNode) {
        if (!shardRouting.started() || !RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shardRouting, allocation))) {
            continue;
        }

        if (allocation.deciders().canRemain(shardRouting, sourceNode, allocation) == Decision.NO) {
            forceMoveShards.add(shardRouting);
        }
    }
});
forceMoveShards.forEach(shard -> tryShardMovementToEligibleNode(eligibleNodes, shard));
```

### PR DESCRIPTION
**Pull request by bugmakerrrrrr**
### Description
Two changes here:
1. Support index level allocation filtering for searchable snapshot index: filter out the shards that cannot remain on the current node and prioritize moving them to eligible nodes before considering moving shards on excluded nodes;
2. Eliminate the logic of swapping primary with replica, as a shard cannot be transitioned from primary to replica mode.
https://github.com/opensearch-project/OpenSearch/blob/6bb53dea24264fc91dc46824041157066a16b74e/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L604

### Related Issues
Resolves #11468

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kotwanikunal** [comment]
> Eliminate the logic of swapping primary with replica, as a shard cannot be transitioned from primary to replica mode
@bugmakerrrrrr In a writable, local index scenario - the swap from replica to primary is not possible. 
For a remote, read-only shard, this is possible in certain scenarios as we can "mark" the shard as primary or replica with a minor performance hit.

**Comment by bugmakerrrrrr** [comment]
> > Eliminate the logic of swapping primary with replica, as a shard cannot be transitioned from primary to replica mode
> > @bugmakerrrrrr In a writable, local index scenario - the swap from replica to primary is not possible.
> > For a remote, read-only shard, this is possible in certain scenarios as we can "mark" the shard as primary or replica with a minor performance hit.

@kotwanikunal I know what you mean, but AFAIU, the cost of relocating a searchable snapshot index is very low, the only side effect is that target shard will experience a cold start. Swapping shards does indeed preserve the cache of the source shard. However, because the default search preference now is `_primary`, swapping shards has very limited effect on improving cold start.

Moreover, in the current implementation, swapping shards can cause the `IllegalArgumentException` (`illegal state: trying to move shard from primary mode to replica mode`). We can certainly skip specific checks for the searchable snapshot index, but I am not sure if this will cause other problems and make the code messy, so I tend to remove the logic of swapping shards.

**Comment by bugmakerrrrrr** [comment]
@kotwanikunal Could you please have a look?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 1]
```diff
@@ -84,31 +84,50 @@ void moveShards() {
...
+                    return;
                 }
 
                 tryShardMovementToEligibleNode(eligibleNodes, ineligibleShard);
             }
         }
     }
 
+    private boolean ineligibleForMove(ShardRouting shard) {
+        return !shard.started() || !RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shard, allocation));
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:128]: The prevailing style is `shard.started() == false || RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shard, allocation) == false`

**Review by kotwanikunal — CHANGES_REQUESTED** [review]
Rest LGTM.

**Comment by bugmakerrrrrr** [comment]
> Rest LGTM.

@kotwanikunal all relevant issues fixed

**Comment by andrross** [comment]
I just pushed a rebase from main and fixed the conflict in the changelog

**Review by kotwanikunal — APPROVED** [review]

**Pull request merged**